numpy/core/src/multiarray/iterators.c

```diff
@@ -501,9 +515,10 @@ iter_subscript_int(PyArrayIterObject *self, PyArrayObject *ind)
         }
     }
 
-    Py_INCREF(PyArray_DESCR(self->ao));
+    PyArray_Descr *dtype = PyArray_DESCR(self->ao);
+    Py_INCREF(dtype);
     ret = (PyArrayObject *)PyArray_NewFromDescr(Py_TYPE(self->ao),
-                             PyArray_DESCR(self->ao),
+                             dtype,
```

Comment:
Do you want to check if this works for `"S0"`m OTOH, its hard to actually get an S0 array to begin with.  Maybe the fix with that is to pull the plug and allow "S0" properly.

Reply:
Yeah, I guess I'm not sure how to get an SO array:

```
In [2]: arr = np.zeros((3, 3), dtype="S0")

In [3]: arr
Out[3]: 
array([[b'', b'', b''],
       [b'', b'', b''],
       [b'', b'', b'']], dtype='|S1')

In [4]: arr.astype("S0")
Out[4]: 
array([[b'', b'', b''],
       [b'', b'', b''],
       [b'', b'', b'']], dtype='|S1')

In [5]: arr.view("S0")
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[5], line 1
----> 1 arr.view("S0")

ValueError: When changing to a smaller dtype, its size must be a divisor of the size of original dtype
```

Reply:
Thanks, let's not worry about it, I guess the only normal way to get it is as a temporary value in `np.asarray([""])`.